torchvision/transforms/autoaugment.py

```diff
@@ -0,0 +1,236 @@
+            (("Equalize", 0.6, None), ("Solarize", 0.6, 6)),
+            (("Invert", 0.9, None), ("Equalize", 0.6, None)),
+            (("Equalize", 0.6, None), ("Rotate", 0.9, 3)),
+            (("ShearX", 0.9, 4), ("AutoContrast", 0.8, None)),
+            (("ShearY", 0.9, 8), ("Invert", 0.4, None)),
+            (("ShearY", 0.9, 5), ("Solarize", 0.2, 6)),
+            (("Invert", 0.9, None), ("AutoContrast", 0.8, None)),
+            (("Equalize", 0.6, None), ("Rotate", 0.9, 3)),
+            (("ShearX", 0.9, 4), ("Solarize", 0.3, 3)),
+            (("ShearY", 0.8, 8), ("Invert", 0.7, None)),
+            (("Equalize", 0.9, None), ("TranslateY", 0.6, 6)),
+            (("Invert", 0.9, None), ("Equalize", 0.6, None)),
+            (("Contrast", 0.3, 3), ("Rotate", 0.8, 4)),
+            (("Invert", 0.8, None), ("TranslateY", 0.0, 2)),
+            (("ShearY", 0.7, 6), ("Solarize", 0.4, 8)),
+            (("Invert", 0.6, None), ("Rotate", 0.8, 4)),
+            (("ShearY", 0.3, 7), ("TranslateX", 0.9, 3)),
+            (("ShearX", 0.1, 6), ("Invert", 0.6, None)),
+            (("Solarize", 0.7, 2), ("TranslateY", 0.6, 7)),
+            (("ShearY", 0.8, 4), ("Invert", 0.8, None)),
+            (("ShearX", 0.7, 9), ("TranslateY", 0.8, 3)),
+            (("ShearY", 0.8, 5), ("AutoContrast", 0.7, None)),
+            (("ShearX", 0.7, 2), ("Invert", 0.1, None)),
+        ]
+
+
+class AutoAugment(torch.nn.Module):
+    r"""AutoAugment data augmentation method based on
+    `"AutoAugment: Learning Augmentation Strategies from Data" <https://arxiv.org/pdf/1805.09501.pdf>`_.
+    The image can be a PIL Image or a Tensor, in which case it is expected
+    to have [..., H, W] shape, where ... means an arbitrary number of leading dimensions.
+
+    Args:
+        policy (AutoAugmentPolicy): Desired policy enum defined by
+            :class:`torchvision.transforms.autoaugment.AutoAugmentPolicy`. Default is ``AutoAugmentPolicy.IMAGENET``.
+        fill (sequence or int or float, optional): Pixel fill value for the area outside the transformed
+            image. If int or float, the value is used for all bands respectively.
+            This option is supported for PIL image and Tensor inputs.
+            If input is PIL Image, the options is only available for ``Pillow>=5.0.0``.
+
+    Example:
+        >>> t = transforms.AutoAugment()
+        >>> transformed = t(image)
+
+        >>> transform=transforms.Compose([
+        >>>     transforms.Resize(256),
+        >>>     transforms.AutoAugment(),
+        >>>     transforms.ToTensor()])
+    """
+
+    def __init__(self, policy: AutoAugmentPolicy = AutoAugmentPolicy.IMAGENET, fill: Optional[List[float]] = None):
+        super().__init__()
+        self.policy = policy
+        self.fill = fill
+        self.transforms = _get_transforms(policy)
+        if self.transforms is None:
+            raise ValueError("The provided policy {} is not recognized.".format(policy))
+
+        _BINS = 10
```

Comment:
Maybe, this can be a configuration even if in the paper/code it is fixed ?

My point is that if folks could play around the functionallity and extend it, it is better than a fixed one ...

Reply:
The AutoAugment policies are fixed per dataset and usually people use it as-is. Although one theoretically could reproduce the paper and learn a policy on a new Dataset, this is unlikely to be be economically viable for most users. That's why my implementation does not focus on extensibility (custom policies/Datasets) and puts all these variables hidden in the constructor of the object. 

To address the immediate comment here, the problem is that the number of bins is related to the search algorithm that AutoAugment used. Since on their paper they encode the policy by referring to the index of the magnitude range (see final pages on the paper), I thought it would be a good idea to implement the code exactly as described on paper. Given the above encoding, one can't just change the number of bins without updating the number of indexes. Doing so will also mean they don't use the AutoAugment policy recommended by the paper anymore but something else. Finally the reason why I use a constant here is to indicate that we use 10 bins, unlike some implementations that use 11.

I start thinking it might be a good idea to revert the change and put back the declarations in the constructor to keep these details hidden... Thoughts?

Reply:
OK, I think we can keep it as it is. Extension of AA is another point and can be addressed in another issue/PR. I was just thinking that as AA was a step in research that led to other learnable dataaugs algorithms, it would be better to have something flexible here...

Reply:
Instead of reverting, I went for the middle ground solution where I move the declarations outside of the constructor to keep things readable on the object and move this to a private method so that they are still fixed. Let me know what you think.

Reply:
Looks good :)